tests/test_ctree.py

```diff
@@ -1,52 +1,57 @@
 
-from diverse_seq.cluster import dvs_ctree
+from diverse_seq.cluster import dvs_ctree, dvs_par_ctree
 
 
-@pytest.mark.parametrize("sketch_size", [400, 4e50])
-def test_ctree_mash(DATA_DIR, sketch_size):
-    dvseqs_path = DATA_DIR / "brca1.dvseqs"
-    app = dvs_ctree(seq_store=dvseqs_path, k=16, sketch_size=sketch_size)
-    tree = app(["Human", "Chimpanzee", "Rhesus", "Horse"])
+def check_ctree_app(app: dvs_ctree | dvs_par_ctree, seqs: SequenceCollection) -> None:
+    tree = app(seqs.take_seqs(["Human", "Chimpanzee", "Rhesus", "Horse"]))
     expected = make_tree("(((Human, Chimpanzee), Rhesus), Horse);")
     assert tree.same_topology(expected)
 
-    tree = app(["Human", "Chimpanzee", "Manatee", "Dugong"])
+    tree = app(seqs.take_seqs(["Human", "Chimpanzee", "Manatee", "Dugong"]))
     expected = make_tree("((Human, Chimpanzee), (Manatee, Dugong));")
     assert tree.same_topology(expected)
 
-    tree = app(["Human", "Chimpanzee", "Manatee", "Dugong", "Rhesus"])
+    tree = app(seqs.take_seqs(["Human", "Chimpanzee", "Manatee", "Dugong", "Rhesus"]))
     expected = make_tree("(((Human, Chimpanzee), Rhesus), (Manatee, Dugong));")
     assert tree.same_topology(expected)
 
 
-def test_ctree_mash_parallel(DATA_DIR):
-    dvseqs_path = DATA_DIR / "brca1.dvseqs"
-    app = dvs_ctree(
-        seq_store=dvseqs_path,
-        k=16,
-        sketch_size=400,
-        numprocs=4,
-    )
-
-    tree = app(["Human", "Chimpanzee", "Manatee", "Dugong", "Rhesus"])
-    expected = make_tree("(((Human, Chimpanzee), Rhesus), (Manatee, Dugong));")
-    assert tree.same_topology(expected)
+@pytest.mark.parametrize("sketch_size", [400, 4e50])
+def test_ctree_mash(unaligned_seqs: SequenceCollection, sketch_size: int) -> None:
+    app = dvs_ctree(k=16, sketch_size=sketch_size, distance_mode="mash")
+    check_ctree_app(app, unaligned_seqs)
 
 
-def test_ctree_euclidean(DATA_DIR):
-    dvseqs_path = DATA_DIR / "brca1.dvseqs"
-    app = dvs_ctree(seq_store=dvseqs_path, k=5, distance_mode="euclidean")
-    tree = app(["Human", "Chimpanzee", "Rhesus", "Horse"])
-    expected = make_tree("(((Human, Chimpanzee), Rhesus), Horse);")
-    assert tree.same_topology(expected)
+def test_ctree_euclidean(unaligned_seqs: SequenceCollection) -> None:
+    app = dvs_ctree(k=5, distance_mode="euclidean")
+    check_ctree_app(app, unaligned_seqs)
 
-    tree = app(["Human", "Chimpanzee", "Manatee", "Dugong"])
-    expected = make_tree("((Human, Chimpanzee), (Manatee, Dugong));")
-    assert tree.same_topology(expected)
 
-    tree = app(["Human", "Chimpanzee", "Manatee", "Dugong", "Rhesus"])
-    expected = make_tree("(((Human, Chimpanzee), Rhesus), (Manatee, Dugong));")
-    assert tree.same_topology(expected)
+@pytest.mark.parametrize("numprocs", [1, 4])
+def test_ctree_mash_parallel(
+    unaligned_seqs: SequenceCollection,
+    numprocs: int,
+) -> None:
+    app = dvs_par_ctree(
+        k=16,
+        sketch_size=400,
+        numprocs=numprocs,
```

Comment:
this is a question, not a critique. The works on github actions don't have 4 cores, does this code adapt the numprocs to be the `min(numprocs, numcpus)`?

Reply:
The Ubuntu/Windows runners according to [this](https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories) have four processors each, and the latest macOS runner has three (I don't mind overspecifying the number of processes here as part of these tests are ensuring the data is gathered back correctly).

The code does not adapt the numprocs to be limited by the number of available cpus (can if you want). Now that I'm thinking about it, should probably allow -1 to use all available cores.

Reply:
Or define a new variable `parallel: bool` to control this and rename `numprocs` to `max_workers`. It's more likely a user just wants `parallel=True`. This former matches the cogent3 app `as_completed()` API. At present that API requires `max_workers` be set via a `par_kw` dict of options for parallelisation (which can also include options such as chunksize).